docs/how-to/incremental-migrations.md

```diff
@@ -0,0 +1,365 @@
+
+```rust
+fn find_character_for_player(ctx: &ReducerContext) -> CharacterV2 {
+    if let Some(character) = ctx.db.character_v2().player_id().find(ctx.sender) {
+        // Already migrated; just return the new player.
+        return character;
+    }
+
+    // Not yet migrated; look up an old character and update it.
+    let old_character = ctx
+        .db
+        .character()
+        .player_id()
+        .find(ctx.sender)
+        .expect("Player has not created a character");
+    ctx.db.character_v2().insert(CharacterV2 {
+        player_id: old_character.player_id,
+        nickname: old_character.nickname,
+        level: old_character.level,
+        class: old_character.class,
+        alliance: Alliance::Neutral,
+    })
+}
+```
+
+Just like when creating a new character, when we update a `character_v2` row, we'll also update the old `character` row, so that outdated clients can continue to function.
+
+```rust
+fn update_character(ctx: &ReducerContext, character: CharacterV2) {
+    ctx.db.character().player_id().update(Character {
+        player_id: character.player_id,
+        nickname: character.nickname.clone(),
+        level: character.level,
+        class: character.class,
+    });
+    ctx.db.character_v2().player_id().update(character);
+}
+```
+
+Then we can make trivial modifications to the callers of `update_character` so that they pass in `CharacterV2` instances:
+
+```rust
+#[spacetimedb::reducer]
+fn rename_character(ctx: &ReducerContext, new_name: String) {
+    let character = find_character_for_player(ctx);
+    log::info!(
+        "Renaming {} to {}",
+        character.nickname,
+        new_name,
+    );
+    update_character(
+        ctx,
+        CharacterV2 {
+            nickname: new_name,
+            ..character
+        },
+    );
+}
+
+#[spacetimedb::reducer]
+fn level_up_character(ctx: &ReducerContext) {
+    let character = find_character_for_player(ctx);
+    log::info!(
+        "Leveling up {} from {} to {}",
+        character.nickname,
+        character.level,
+        character.level + 1,
+    );
+    update_character(
+        ctx,
+        CharacterV2 {
+            level: character.level + 1,
+            ..character
+        },
+    );
+}
+```
+
+And finally, we can define our new `choose_alliance` reducer:
+
+```rust
+#[spacetimedb::reducer]
+fn choose_alliance(ctx: &ReducerContext, alliance: Alliance) {
+    let character = find_character_for_player(ctx);
+    log::info!(
+        "Setting alliance of {} to {:?}",
+        character.nickname,
+        alliance,
+    );
+    update_character(
+        ctx,
+        CharacterV2 {
+            alliance,
+            ..character
+        },
+    );
+}
+```
+
```

Comment:
Database developers will also want to add a scheduled reducer that checks that the two tables are in sync.

Reply:
This seems like an anti-pattern and also like a great way to write stupidly slow reducers that block your whole game.

Reply:
Yeah but having out-of-sync tables is a recipe for item duplication bugs and the like.

I guess you could do such a query manually with the CLI...

Reply:
Another solution would be:
- Write code that asserts the integrity of synced tables for a particular row ID.
- Call this after any write to synced tables, in particular after any initialization or update code.

This will be mildly slow but I think it's a small price to pay to guarantee database integrity.

Reply:
I guess `update_character` suffices for enforcement of this. See my other comment above.